components/dashboard/src/components/RepositoryFinder.tsx

```diff
@@ -0,0 +1,171 @@
+    const [searchResults, setSearchResults] = useState<SearchResult[]>([]);
+    const [selectedSearchResult, setSelectedSearchResult] = useState<SearchResult | undefined>();
+
+    const onResults = (results: SearchResult[]) => {
+        if (JSON.stringify(results) !== JSON.stringify(searchResults)) {
+            setSearchResults(results);
+            setSelectedSearchResult(results[0]);
+        }
+    }
+
+    const search = async (query: string) => {
+        setSearchQuery(query);
+        await findResults(query, onResults);
+        if (await refreshSearchData(query, user)) {
+            // Re-run search if the underlying search data has changed
+            await findResults(query, onResults);
+        }
+    }
+
+    // Up/Down keyboard navigation between results
+    const onKeyDown = (event: React.KeyboardEvent) => {
+        if (!selectedSearchResult) {
+            return;
+        }
+        const selectedIndex = searchResults.indexOf(selectedSearchResult);
+        const select = (index: number) => {
+            // Implement a true modulus in order to "wrap around" (e.g. `select(-1)` should select the last result)
+            // Source: https://stackoverflow.com/a/4467559/3461173
+            const n = Math.min(searchResults.length, MAX_DISPLAYED_ITEMS);
+            setSelectedSearchResult(searchResults[((index % n) + n) % n]);
+        }
+        if (event.key === 'ArrowDown') {
+            event.preventDefault();
+            select(selectedIndex + 1);
+            return;
+        }
+        if (event.key === 'ArrowUp') {
+            event.preventDefault();
+            select(selectedIndex - 1);
+            return;
+        }
+    }
+
+    useEffect(() => {
+        const element = document.querySelector(`a[href='/#${selectedSearchResult}']`);
+        if (element) {
+            element.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
+        }
+    }, [selectedSearchResult]);
+
+    const onSubmit = (event: React.FormEvent) => {
+        event.preventDefault();
+        if (selectedSearchResult) {
+            window.location.href = '/#' + selectedSearchResult;
+        }
+    }
+
+    return <form onSubmit={onSubmit}>
+        <div className="flex px-4 rounded-xl border border-gray-300 dark:border-gray-500">
+            <div className="py-4">
+                <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 16 16" width="16" height="16"><path fill="#A8A29E" d="M6 2a4 4 0 100 8 4 4 0 000-8zM0 6a6 6 0 1110.89 3.477l4.817 4.816a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 010 6z" /></svg>
+            </div>
+            <input type="search" className="flex-grow" placeholder="repository" autoFocus value={searchQuery} onChange={e => search(e.target.value)} onKeyDown={onKeyDown} />
+        </div>
+        <div className="mt-3 -mx-5 px-5 flex flex-col space-y-2 h-64 overflow-y-auto">
+            {searchQuery === '' && searchResults.length === 0 &&
```

Comment:
**issue**: When you search for something and then delete the text input you are still seeing some arbitrary results below. Expected?

**suggestion**: What do you think of rolling back to the emopty (default) state when deleting the search input?

Reply:
This is sort of an edge case, and I don't know what would be the best UX here.

I've implemented it like this because:
- The initial empty state teaches you about the feature, but once you've started typing I considered you've moved on from "what is this?" to "how does it work?" (hence no need for the empty state text anymore)
- I found it marginally helpful to know what results were available _without_ any filter: the current state allows me to know how many possible results there are, and to get an idea of what they might look like (e.g. "aha, these look like repositories -- some I've contributed to, and some examples")

Going back to the empty state could work, but this makes the "what are all possible results" edge use case harder to achieve (e.g. you'll need to type a common character like `h` or `/`, which is less discoverable).

Anyway, no strong opinion. Do you much prefer restoring the empty state on empty input?

Reply:
Thanks for making the call! 🤙 Will adjust this.